In [50]:
import numpy as np
import pandas as pd

In [51]:
# load dataset in a dataframe
df_titanic = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/refs/heads/master/titanic.csv")

In [52]:
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


In [53]:
df_titanic.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [54]:
# Drop any variable that is missing more than 70%
df_titanic.drop('deck', axis=1,inplace=True)

In [55]:
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  embark_town  889 non-null    object 
 12  alive        891 non-null    object 
 13  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(6)
memory usage: 85.4+ KB


In [56]:
df_titanic.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True


In [57]:
# Display thos record which contain Nan value in colum embark_town
df_titanic[df_titanic['embark_town'].isnull()]

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
61,1,1,female,38.0,0,0,80.0,NaN,First,woman,False,NaN,yes,True
829,1,1,female,62.0,0,0,80.0,NaN,First,woman,False,NaN,yes,True


In [58]:
(2/891)*100

0.22446689113355783

In [59]:
# As there are only 2 record with missing embark_town, so we can delete them safely.
df_titanic.drop([61, 829], axis=0, inplace=True)
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
Index: 889 entries, 0 to 890
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     889 non-null    int64  
 1   pclass       889 non-null    int64  
 2   sex          889 non-null    object 
 3   age          712 non-null    float64
 4   sibsp        889 non-null    int64  
 5   parch        889 non-null    int64  
 6   fare         889 non-null    float64
 7   embarked     889 non-null    object 
 8   class        889 non-null    object 
 9   who          889 non-null    object 
 10  adult_male   889 non-null    bool   
 11  embark_town  889 non-null    object 
 12  alive        889 non-null    object 
 13  alone        889 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(6)
memory usage: 92.0+ KB


In [60]:
((889-712)/889)*100

19.910011248593925

In [61]:
df_titanic.head(2)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False


In [62]:
df_titanic.drop(['sex','embarked','class','who','embark_town','alive'], axis=1).corr()

,survived,pclass,age,sibsp,parch,fare,adult_male,alone
survived,1.000000,-0.335549,-0.082446,-0.034040,0.083151,0.255290,-0.555520,-0.206207
pclass,-0.335549,1.000000,-0.365902,0.081656,0.016824,-0.548193,0.090098,0.138553
age,-0.082446,-0.365902,1.000000,-0.307351,-0.187896,0.093143,0.286543,0.195766
sibsp,-0.034040,0.081656,-0.307351,1.000000,0.414542,0.160887,-0.255401,-0.584186
parch,0.083151,0.016824,-0.187896,0.414542,1.000000,0.217532,-0.351945,-0.583112
fare,0.255290,-0.548193,0.093143,0.160887,0.217532,1.000000,-0.179851,-0.274079
adult_male,-0.555520,0.090098,0.286543,-0.255401,-0.351945,-0.179851,1.000000,0.407992
alone,-0.206207,0.138553,0.195766,-0.584186,-0.583112,-0.274079,0.407992,1.000000


In [63]:
# Identify how many unique value available
df_titanic['pclass'].unique()

array([3, 1, 2])

In [64]:
# Calculate average of age for each class
df_titanic.groupby('pclass')['age'].mean()

,age
pclass,
1,38.105543
2,29.877630
3,25.140620


In [65]:
# Custom function to return age depending on pclass
def impute_age(cols):
  age = cols[0]
  pclass = cols[1]
  if pd.isnull(age):
    if pclass==1:
      return 38
    elif pclass==2:
      return 29
    else:
      return 25
  else:
    return age

In [66]:
df_titanic['age'] = df_titanic[['age','pclass']].apply(impute_age,axis=1)

/tmp/ipykernel_2488/4267891560.py:3: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  age = cols[0]
/tmp/ipykernel_2488/4267891560.py:4: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  pclass = cols[1]


In [67]:
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
Index: 889 entries, 0 to 890
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     889 non-null    int64  
 1   pclass       889 non-null    int64  
 2   sex          889 non-null    object 
 3   age          889 non-null    float64
 4   sibsp        889 non-null    int64  
 5   parch        889 non-null    int64  
 6   fare         889 non-null    float64
 7   embarked     889 non-null    object 
 8   class        889 non-null    object 
 9   who          889 non-null    object 
 10  adult_male   889 non-null    bool   
 11  embark_town  889 non-null    object 
 12  alive        889 non-null    object 
 13  alone        889 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(6)
memory usage: 92.0+ KB


In [68]:
df_titanic.groupby('pclass')['age'].mean()

,age
pclass,
1,38.090748
2,29.825163
3,25.101670


In [69]:
df_titanic.head(5)

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True


In [70]:
# In DF, we need to find potential column with non-numeric values

df_titanic['sex'].unique()

array(['male', 'female'], dtype=object)

In [71]:
pd.get_dummies(df_titanic['sex']).head() # without drop_first

,female,male
0,False,True
1,True,False
2,True,False
3,True,False
4,False,True


In [72]:
df_titanic['sex'].head()

,sex
0,male
1,female
2,female
3,female
4,male


In [73]:
pd.get_dummies(df_titanic['sex'],drop_first=True).head()

,male
0,True
1,False
2,False
3,False
4,True


In [74]:
gender = pd.get_dummies(df_titanic['sex'],drop_first=True)
df_titanic.drop('sex',axis=1, inplace = True)

In [75]:
df_titanic = pd.concat([df_titanic, gender], axis=1)
df_titanic.head()

,survived,pclass,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone,male
0,0,3,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False,True
1,1,1,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False,False
2,1,3,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True,False
3,1,1,35.0,1,0,53.1000,S,First,woman,False,Southampton,yes,False,False
4,0,3,35.0,0,0,8.0500,S,Third,man,True,Southampton,no,True,True


In [76]:
df_titanic.head(3)

,survived,pclass,age,sibsp,parch,fare,embarked,class,who,adult_male,embark_town,alive,alone,male
0,0,3,22.0,1,0,7.2500,S,Third,man,True,Southampton,no,False,True
1,1,1,38.0,1,0,71.2833,C,First,woman,False,Cherbourg,yes,False,False
2,1,3,26.0,0,0,7.9250,S,Third,woman,False,Southampton,yes,True,False


In [77]:
df_titanic['embarked'].unique()
# embarked - ['S', 'C', 'Q']

array(['S', 'C', 'Q'], dtype=object)

In [78]:
df_titanic['class'].unique()
# class - ['Third', 'First', 'Second']

array(['Third', 'First', 'Second'], dtype=object)

In [79]:
df_titanic['who'].unique()
# child - ['man', 'woman', 'child']

array(['man', 'woman', 'child'], dtype=object)

In [80]:
df_titanic['alive'].unique()
# alive - ['no', 'yes']

array(['no', 'yes'], dtype=object)

In [81]:
df_titanic.drop(['embarked','class','alive'], axis=1, inplace= True )

In [82]:
df_titanic.head(3)

,survived,pclass,age,sibsp,parch,fare,who,adult_male,embark_town,alone,male
0,0,3,22.0,1,0,7.2500,man,True,Southampton,False,True
1,1,1,38.0,1,0,71.2833,woman,False,Cherbourg,False,False
2,1,3,26.0,0,0,7.9250,woman,False,Southampton,True,False


In [83]:
who = pd.get_dummies(df_titanic['who'])
df_titanic.drop('who', axis=1, inplace=True)

df_titanic = pd.concat([df_titanic, who], axis =1 )

In [84]:
df_titanic.head(3)

,survived,pclass,age,sibsp,parch,fare,adult_male,embark_town,alone,male,child,man,woman
0,0,3,22.0,1,0,7.2500,True,Southampton,False,True,False,True,False
1,1,1,38.0,1,0,71.2833,False,Cherbourg,False,False,False,False,True
2,1,3,26.0,0,0,7.9250,False,Southampton,True,False,False,False,True


In [85]:
embark_town = pd.get_dummies(df_titanic['embark_town'])
df_titanic.drop('embark_town', axis=1, inplace=True)

df_titanic = pd.concat([df_titanic, embark_town], axis =1 )

In [86]:
df_titanic.head(3)

,survived,pclass,age,sibsp,parch,fare,adult_male,alone,male,child,man,woman,Cherbourg,Queenstown,Southampton
0,0,3,22.0,1,0,7.2500,True,False,True,False,True,False,False,False,True
1,1,1,38.0,1,0,71.2833,False,False,False,False,False,True,True,False,False
2,1,3,26.0,0,0,7.9250,False,True,False,False,False,True,False,False,True


In [87]:
df_titanic.info()

<class 'pandas.core.frame.DataFrame'>
Index: 889 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     889 non-null    int64  
 1   pclass       889 non-null    int64  
 2   age          889 non-null    float64
 3   sibsp        889 non-null    int64  
 4   parch        889 non-null    int64  
 5   fare         889 non-null    float64
 6   adult_male   889 non-null    bool   
 7   alone        889 non-null    bool   
 8   male         889 non-null    bool   
 9   child        889 non-null    bool   
 10  man          889 non-null    bool   
 11  woman        889 non-null    bool   
 12  Cherbourg    889 non-null    bool   
 13  Queenstown   889 non-null    bool   
 14  Southampton  889 non-null    bool   
dtypes: bool(9), float64(2), int64(4)
memory usage: 56.4 KB


In [88]:
df_titanic.corr()

,survived,pclass,age,sibsp,parch,fare,adult_male,alone,male,child,man,woman,Cherbourg,Queenstown,Southampton
survived,1.000000,-0.335549,-0.054899,-0.034040,0.083151,0.255290,-0.555520,-0.206207,-0.541585,0.137284,-0.555520,0.504460,0.169966,0.004536,-0.151777
pclass,-0.335549,1.000000,-0.400979,0.081656,0.016824,-0.548193,0.090098,0.138553,0.127741,0.121140,0.090098,-0.172640,-0.245733,0.220558,0.076466
age,-0.054899,-0.400979,1.000000,-0.242393,-0.173838,0.118665,0.249012,0.167452,0.085232,-0.556423,0.249012,0.087304,0.043699,-0.070862,0.006319
sibsp,-0.034040,0.081656,-0.242393,1.000000,0.414542,0.160887,-0.255401,-0.584186,-0.116348,0.352224,-0.255401,0.048824,-0.060074,-0.026692,0.069438
parch,0.083151,0.016824,-0.173838,0.414542,1.000000,0.217532,-0.351945,-0.583112,-0.247508,0.351268,-0.351945,0.152210,-0.011588,-0.081585,0.061512
fare,0.255290,-0.548193,0.118665,0.160887,0.217532,1.000000,-0.179851,-0.274079,-0.179958,0.004452,-0.179851,0.188650,0.270731,-0.116684,-0.163758
adult_male,-0.555520,0.090098,0.249012,-0.255401,-0.351945,-0.179851,1.000000,0.407992,0.908250,-0.396358,1.000000,-0.813572,-0.067449,-0.077783,0.108062
alone,-0.206207,0.138553,0.167452,-0.584186,-0.583112,-0.274079,0.407992,1.000000,0.306985,-0.347112,0.407992,-0.214510,-0.094513,0.087099,0.027982
male,-0.541585,0.127741,0.085232,-0.116348,-0.247508,-0.179958,0.908250,0.306985,1.000000,-0.112364,0.908250,-0.895758,-0.084520,-0.075217,0.121405
child,0.137284,0.121140,-0.556423,0.352224,0.351268,0.004452,-0.396358,-0.347112,-0.112364,1.000000,-0.396358,-0.211374,0.022862,-0.030089,-0.001090


A correlation matrix from the Titanic dataset. Correlation values range from -1 to +1:

+1 = Strong positive relationship
-1 = Strong negative relationship
0 = No linear relationship

# Key Factors Affecting Survival
Gender had the biggest impact
Feature	Correlation with Survived
male	-0.542
woman	+0.504
adult_male	-0.556

Interpretation
Male passengers were much less likely to survive.
Women were much more likely to survive.
This supports the historical "Women and children first" evacuation policy.

Business Insight:
Gender was one of the strongest predictors of survival.

Passenger Class Matters
Feature	Correlation
pclass	-0.336
Interpretation

#Higher class number means lower class:

1 = First Class
2 = Second Class
3 = Third Class

Negative correlation means:

Higher class number
↓
Lower survival

So:

First Class > Second Class > Third Class

in survival rate.

Insight:
#Wealthier passengers had better survival chances.

Ticket Fare
Feature	Correlation
fare	+0.255
Interpretation

#Higher fare passengers survived more often.

This aligns with:

High Fare
↓
First Class
↓
Better Survival
Traveling Alone
Feature	Correlation
alone	-0.206
Interpretation

Passengers traveling alone had lower survival rates.

#People traveling with family may have:

Received assistance
Stayed together
Been prioritized during evacuation
Children
Feature	Correlation
child	+0.137
Interpretation

Children had slightly higher survival rates.

This again supports:

#Women and Children First
2. Weak Factors
Age
Feature	Correlation
age	-0.055
Interpretation

Age alone has almost no direct relationship with survival.

Older people were only slightly less likely to survive.

#Number of Siblings/Spouses
Feature	Correlation
sibsp	-0.034

Very weak relationship.

#Number of Parents/Children
Feature	Correlation
parch	+0.083

Very weak positive relationship.

3. Port of Embarkation Analysis
Cherbourg
Feature	Correlation
Cherbourg	+0.170

#Passengers boarding from Cherbourg survived more often.

Southampton
Feature	Correlation
Southampton	-0.152

#Passengers boarding from Southampton survived less often.

Queenstown
Feature	Correlation
Queenstown	+0.005

No meaningful impact.

4. Relationships Between Independent Variables
#Fare vs Passenger Class
Variables	Correlation
fare vs pclass	-0.548
Interpretation

As passenger class number increases:

1st Class → Expensive
3rd Class → Cheap

Fare decreases.

This is expected.

#Child vs Age
Variables	Correlation
age vs child	-0.557

Strong negative correlation.

Older passengers are obviously less likely to be classified as children.

#Male vs Woman
Variables	Correlation
male vs woman	-0.896

Very strong negative correlation.

A passenger cannot generally be both.

#Alone vs Family
Variables	Correlation
alone vs sibsp	-0.584
alone vs parch	-0.583

Passengers traveling alone naturally have fewer family members aboard.

# Executive Summary
Strongest Positive Survival Factors
Feature	Correlation
woman	+0.504
fare	+0.255
Cherbourg	+0.170
child	+0.137
Strongest Negative Survival Factors
Feature	Correlation
adult_male	-0.556
male	-0.542
pclass	-0.336
alone	-0.206
Southampton	-0.152

In [90]:
df_titanic.head(4)

,survived,pclass,age,sibsp,parch,fare,adult_male,alone,male,child,man,woman,Cherbourg,Queenstown,Southampton
0,0,3,22.0,1,0,7.2500,True,False,True,False,True,False,False,False,True
1,1,1,38.0,1,0,71.2833,False,False,False,False,False,True,True,False,False
2,1,3,26.0,0,0,7.9250,False,True,False,False,False,True,False,False,True
3,1,1,35.0,1,0,53.1000,False,False,False,False,False,True,False,False,True


In [96]:
df_titanic[df_titanic['woman'] == True]

,survived,pclass,age,sibsp,parch,fare,adult_male,alone,male,child,man,woman,Cherbourg,Queenstown,Southampton
1,1,1,38.0,1,0,71.2833,False,False,False,False,False,True,True,False,False
2,1,3,26.0,0,0,7.9250,False,True,False,False,False,True,False,False,True
3,1,1,35.0,1,0,53.1000,False,False,False,False,False,True,False,False,True
8,1,3,27.0,0,2,11.1333,False,False,False,False,False,True,False,False,True
11,1,1,58.0,0,0,26.5500,False,True,False,False,False,True,False,False,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
880,1,2,25.0,0,1,26.0000,False,False,False,False,False,True,False,False,True
882,0,3,22.0,0,0,10.5167,False,True,False,False,False,True,False,False,True
885,0,3,39.0,0,5,29.1250,False,False,False,False,False,True,False,True,False
887,1,1,19.0,0,0,30.0000,False,True,False,False,False,True,False,False,True
